### PR TITLE
[AIRFLOW-5391] Do not run skipped tasks when they are cleared

### PR DESCRIPTION
**Pull request by yuqian90**
This PR fixes the following issue:

If a task is skipped by ``BranchPythonOperator``, ``BaseBranchOperator`` orr ``ShortCircuitOperator`` and the user then clears the skipped task, it'll execute. 

After this PR:
The ``NotPreviouslySkippedDep`` rule will first evaluate if a task has a direct ``SkipMixin`` parent that has decided to skip it. This is done by examining the XCom data stored by ``SkipMixin.skip()`` or ``SkipMixin.skip_all_except()``.

The implementation is inspired by the author of [this blog](https://blog.diffractive.io/2018/08/07/replacement-shortcircuitoperator-for-airflow/).

---
Issue link: [AIRFLOW-5391](https://issues.apache.org/jira/browse/AIRFLOW-5391)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by yuqian90** [comment]
This is work-in-progress and untested. I want to get some opinions from committers first to see if it's worth pursuing. I'll then add tests and make the PR complete.

One immediate shortcoming that I can see:
This PR only makes skipped tasks respect the decision of upstream `BranchPythonOperator` when cleared. Other variants of branching exist, namely `BaseBranchOperator` and `ShortCircuitOperator`. This PR does not make tasks respect the decision of those operators when cleared.

**Comment by tooptoop4** [comment]
hmm, I use BranchPythonOperator and clear command a lot and I never see the skipped tasks just run on clear. do u use none_failed ?

**Comment by yuqian90** [comment]
@tooptoop4 the problem happens when you clear the skipped task (not when you clear the BranchPythonOperator). There's an example in the linked JIRA.

**Comment by tooptoop4** [comment]
i clear entire dag and dont face problem

**Comment by yuqian90** [comment]
@tooptoop4 yes if always clear the entire DAG, the BranchPythonOperator itself gets cleared so you won't face this problem.

**Comment by tooptoop4** [comment]
makes sense!

**Comment by yuqian90** [comment]
- Added tests
- Extended the same support to ``BranchPythonOperator``, ``ShortCircuitOperator`` and ``BaseBranchOperator``. Clearing downstream tasks of all these three operators will now do the right thing.
- That said, I still think ``BaseBranchOperator`` is a bit redundant since we already have ``BranchPythonOperator``. I'm proposing to deprecate ``BaesBranchOperator`` in [another PR](https://github.com/apache/airflow/pull/7331)

The PR is now ready for review. Reviewers @kaxil  @ashb  @feluelle please take a look.

**Comment by yuqian90** [comment]
@ashb  @kaxil  @feluelle  @tooptoop4  Hoping to get a review here.

I have updated the PR a bit. The logic is still the same, but the code that does the branching evaluation has been moved into a new dependency rule inside ``ti_deps/deps/branch_dep.py``.

When `BranchDep` is evaluated, it sets the state of a task to "SKIPPED" if it has a parent that decided to skip it.

For example, this comes from "example_short_circuit_operatorschedule":
image

For a DAG that looks like this, `condition_is_False` decided to skip `false_1`. Before this PR, the problem is if someone clears `false_1`, it'll execute. This is very counter-intuitive for users:
image
image


After this PR, when the scheduler evaluates the dependency rules for `false_1`, `BranchDep` will skip `false_1` because it knows that the parent task `condition_is_False` had already decided to skip `false_1`. In other words, this makes the "skipped" status "sticky".
 
So after this PR, the DAG will look like this after someone clears `false_1`. This is much more intuitive:

image

File: airflow/models/skipmixin.py [hunk 1]
```diff
@@ -98,3 +98,13 @@ def skip_all_except(self, ti: TaskInstance, branch_task_ids: Union[str, Iterable
 
             self.log.info("Skipping tasks %s", [t.task_id for t in skip_tasks])
             self.skip(dag_run, ti.execution_date, skip_tasks)
+
+    def evaluate_skip_condition(self, xcom_value, downstream_ti):
```
Comment (feluelle) [inline airflow/models/skipmixin.py]: Please add types to the function parameters.

and also in the docs, please.

File: airflow/models/skipmixin.py [hunk 2]
```diff
@@ -98,3 +98,13 @@ def skip_all_except(self, ti: TaskInstance, branch_task_ids: Union[str, Iterable
...
+        """
+        This method evalutes the xcom_value of the SkipMixin itself and decides if a downstream task
+        should be skipped. Returns True if the downstream_ti should be skipped. Otherwise False.
```
Comment (feluelle) [inline airflow/models/skipmixin.py]: You can add `:returns:` and `:rtype:` to the end of the docstring instead of putting it to the description.

File: tests/operators/test_python.py [hunk 3]
```diff
@@ -541,6 +542,62 @@ def test_xcom_push(self):
...
+        tis = dr.get_task_instances()
+        for ti in tis:
+            if ti.task_id == 'make_choice':
+                self.assertEqual(ti.state, State.SUCCESS)
+            elif ti.task_id == 'branch_1':
+                self.assertEqual(ti.state, State.SUCCESS)
+            elif ti.task_id == 'branch_2':
+                self.assertEqual(ti.state, State.SKIPPED)
+            else:
+                raise Exception
```
Comment (feluelle) [inline tests/operators/test_python.py]:
```suggestion
                raise ValueError(f'Invalid task id {ti.task_id} found!')
```

WDYT?

Comment (feluelle) [reply tests/operators/test_python.py]: and in the other cases as well?

**Review by feluelle — COMMENTED** [review]
Overall LGTM. Great test coverage 👍 

..would be nice if you could complete the docstrings to contain types and add the types also to the function header.

**Comment by yuqian90** [comment]
Hi, @feluelle  I've adopted all your suggestions (adding doc and raise ValueError), and the PR has been updated. Please take another look. Thank you!

File: airflow/ti_deps/deps/branch_dep.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                start_date=ti.execution_date,
+                end_date=ti.execution_date,
+                state=State.finished() + [State.UPSTREAM_FAILED],
+                session=session)
+
+        finished_task_ids = {t.task_id for t in finished_tasks}
+
+        for parent in upstream:
+            if isinstance(parent, SkipMixin):
+                xcom_value = ti.xcom_pull(task_ids=parent.task_id)
```
Comment (ashb) [inline airflow/ti_deps/deps/branch_dep.py]: Waait what. I'm really not sure about this -- I don't see from the problem you've describe why we have to touch XCom at all.

Can you explain this in more detail?

Comment (ashb) [reply airflow/ti_deps/deps/branch_dep.py]: And if we are going to use Xcom for this (which as I say, I'm not convinced) we should use something other than the default key.

Comment (feluelle) [reply airflow/ti_deps/deps/branch_dep.py]: > we should use something other than the default key.

Oh. Ash ist totally right @yuqian90 I missed that. We cannot use the default key.

> Waait what. I'm really not sure about this -- I don't see from the problem you've describe why we have to touch XCom at all.

Why don't you want to use xcom for this?

Comment (yuqian90) [reply airflow/ti_deps/deps/branch_dep.py]: Hi, @ashb  here's the reason I'm using XCom, using the same example I previously provided:

In this example, `condition_false` is a `ShortCircuitOperator` that decided to skip its downstream task. But If the user clicks `false_1` and hit "Clear", what Airflow currently does is that it'll actually execute `false_1`, which is very counter-intuitive for users.

So how do we improve this behavior?

When `false_1` is cleared, one way to know that it should actually be skipped is by looking at what its parent `ShortCircuitOperator` returned. To get the return value of the `ShortCircuitOperator`, I'm resorting to `XCom`.

image


This works fine. But maybe you know a better way than using XCom for this purpose. Definitely let me know I don't mind changing.

Comment (ashb) [reply airflow/ti_deps/deps/branch_dep.py]: Oh - to start with I thought this was Xcom of false_1 etc, not of the skipping operator.

Let me look at this again.

File: airflow/ti_deps/deps/branch_dep.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    def _get_dep_statuses(self, ti, session, dep_context=None):
+        from airflow.models.skipmixin import SkipMixin
+        from airflow.utils.state import State
+
+        upstream = ti.task.get_direct_relatives(upstream=True)
+
+        finished_tasks = dep_context.finished_tasks if dep_context else None
+        if finished_tasks is None:
+            # this is for the strange feature of running tasks without dag_run
+            finished_tasks = ti.task.dag.get_task_instances(
```
Comment (ashb) [inline airflow/ti_deps/deps/branch_dep.py]: This is now the 2nd? 3rd? time this block of code exists, please could you refactor it to a method on dep_context and use it in all the places this is called (Please rebase to master too, as there is a new one since you opened this PR)

Comment (ashb) [reply airflow/ti_deps/deps/branch_dep.py]: This new method should also cache/memoize the result back in to `dep_context.finished_tasks`.

File: airflow/models/skipmixin.py [hunk 6]
```diff
@@ -98,3 +98,17 @@ def skip_all_except(self, ti: TaskInstance, branch_task_ids: Union[str, Iterable
 
             self.log.info("Skipping tasks %s", [t.task_id for t in skip_tasks])
             self.skip(dag_run, ti.execution_date, skip_tasks)
+
+    def evaluate_skip_condition(self, xcom_value: Any, downstream_ti: TaskInstance):
+        """
+        Evalutes the xcom_value of the SkipMixin itself and decides if a downstream task
+        should be skipped.
```
Comment (ashb) [inline airflow/models/skipmixin.py]: This doc needs expanding to say when this is called (as it's not clear without reading the rest of this code) so people with custom operators know what to do.

File: airflow/operators/branch_operator.py [hunk 7]
```diff
@@ -48,4 +48,21 @@ def choose_branch(self, context: Dict) -> Union[str, Iterable[str]]:
...
+        :param downstream_ti: The TaskInstance of child task
+        :type downstream_ti: airflow.models.TaskInstance
+        :return: True if xcom_value indicates downstream_ti should be skipped. Otherwise False.
+        :rtype: bool
+        """
+        branch = xcom_value
+        if isinstance(branch, str):
+            branch = [branch]
+
+        return downstream_ti.task_id not in branch
```
Comment (ashb) [inline airflow/operators/branch_operator.py]: Rather than having this logic here, and needing to implement it each time, what do you think instead of having `skip` and `skip_all_except` methods in the mixin storing the list of tasks it has skipped in to Xcom. That way this evaluate function doesn't need to be subclassed, the parent mix in can handle it for all cases.

That does possibly make us store more data in Xcom, so maybe that's not the best plan.

File: airflow/ti_deps/deps/branch_dep.py (hunk 4 above)
Comment (ashb) [inline airflow/ti_deps/deps/branch_dep.py]: There is no point pulling the xcom  if `parent.task_id not in finished_task_ids` - avoiding extra DB requests in the hot-path of the scheduler is something we want to do.

File: airflow/ti_deps/deps/branch_dep.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    NAME = "Branch Rule"
+    IGNORABLE = True
+    IS_TASK_DEP = True
+
+    def _get_dep_statuses(self, ti, session, dep_context=None):
+        from airflow.models.skipmixin import SkipMixin
+        from airflow.utils.state import State
+
+        upstream = ti.task.get_direct_relatives(upstream=True)
```
Comment (ashb) [inline airflow/ti_deps/deps/branch_dep.py]: In your example, if I clear false_2 it would re run -- the skip task was not a direct relative. That makes this behaviour even more confusing as sometimes it would run and sometimes it wouldn't depending on which task was cleared.

File: tests/operators/test_python.py [hunk 9]
```diff
@@ -396,7 +397,7 @@ def test_without_dag_run(self):
                 elif ti.task_id == 'branch_2':
                     self.assertEqual(ti.state, State.SKIPPED)
                 else:
-                    raise Exception
+                    raise ValueError(f'Invalid task id {ti.task_id} found!')
```
Comment (ashb) [inline tests/operators/test_python.py:400]: What is this change for? It seems unrelated to your change.

File: airflow/models/baseoperator.py [hunk 10]
```diff
@@ -650,6 +651,7 @@ def deps(self) -> Set[BaseTIDep]:
             NotInRetryPeriodDep(),
             PrevDagrunDep(),
             TriggerRuleDep(),
+            BranchDep(),
```
Comment (ashb) [inline airflow/models/baseoperator.py]:
```suggestion
            NotPreviouslySkipped(),
```

There are more ways of skipping than just branches.

**Review by ashb — CHANGES_REQUESTED** [review]
Okay, why you are using XCom makes more sense now but I still think it should use a non-default Key (as skip Mixin could be used elsewhere, and we don't want to overwrite possible xcom from the task)

File: airflow/ti_deps/deps/branch_dep.py (hunk 8 above)
Comment (yuqian90) [reply airflow/ti_deps/deps/branch_dep.py]: I did have the same concern as you. So I did some experiment. If someone clears `false_2`, Airflow actually does the right thing. It becomes skipped. That's already happening in Airflow, not because of my PR.
The doc describes this behaviour as "skipped states propagates". It's actually `trigger_rule_deps.py` that does this "magic". Somewhere here the state is set to SKIPPED if any of its upstream is skipped and the task `trigger_rule` is `all_success`. This is exactly what we hope for.

```LANG=python
            if tr == TR.ALL_SUCCESS:
                if upstream_failed or failed:
                    ti.set_state(State.UPSTREAM_FAILED, session)
                elif skipped:
                    ti.set_state(State.SKIPPED, session)
```

File: tests/operators/test_python.py (hunk 9 above)
Comment (yuqian90) [reply tests/operators/test_python.py:400]: I was changing this from `Exception` to `ValueError` in one of the new test cases I added following previous suggestions from reviewers. So I replaced the same usage in this test module too. I will revert this change if you prefer leaving the existing code untouched.

File: airflow/operators/branch_operator.py (hunk 7 above)
Comment (yuqian90) [reply airflow/operators/branch_operator.py]: That might actually be one way to do this. There's just one downside i can see. 

For `BranchPythonOperator`, the current behaviour in my PR is to store the task_ids to follow (because that's in XCom). 

If we change this to store the list of task_ids skipped in `SkipMixin`, people making incremental changes to their existing DAG may find some surprising behaviour. E.g if the DAG looks like this on 20200101. And the Branch follows A, skips B and C. `SkipMixin` stores B and C in db as its "list of task_ids skipped".
```
Branch >> [A, B, C]
```

We then add a D on 20200102. The DAG becomes like this. 
```
Branch >> [A, B, C, D]
```

Branch itself is not changed. It still only follows A. But if someone clears a task on the 20200101 DAG. It'll start running D because D is not found in the"list of task_ids skipped" of the `SkipMixin`.

I think this behaviour is less intuitive than having Branch automatically skip anything it did not follow (like what this PR is currently doing).

However I do like your suggestion because the implementation is simpler and more elegant than forcing developers to implement an additional `evaluate_skip_condition()`.

What do you think?

Comment (yuqian90) [reply airflow/operators/branch_operator.py]: @ashb  how about this:

Since there's `skip()` and `skip_all_except()`. I can make `skip()` store what's skipped, and make `skip_all_except()` store what's in the "except". 
So the XCom data will look something along this line:
```
"skipped": [B, C]
"except": [A]
```

When looking up this XCom values, first, skip [B, C], and then also skip everything that's not equal in "except".

That should cover all the edge cases that can happen.

File: airflow/models/baseoperator.py (hunk 10 above)
Comment (yuqian90) [reply airflow/models/baseoperator.py]: @ashb  what about calling it `PreviouslySkipped` instead of `NotPreviouslySkipped`?

I'm just worried the negation in the naming can be a little confusion.

File: airflow/ti_deps/deps/branch_dep.py (hunk 8 above)
Comment (ashb) [reply airflow/ti_deps/deps/branch_dep.py]: Ah great! I wonder if that behaviour is tested already or not

File: tests/operators/test_python.py (hunk 9 above)
Comment (ashb) [reply tests/operators/test_python.py:400]: Cool keep it

File: airflow/models/baseoperator.py (hunk 10 above)
Comment (ashb) [reply airflow/models/baseoperator.py]: We have `NotInRetryPeriodDep` already , so `NotPreviouslySkippedDep` would follow -- `PreviouslySkippedDep` would _pass_ if it had been skipped, the opposite of what we want.

File: airflow/ti_deps/deps/branch_dep.py (hunk 8 above)
Comment (yuqian90) [reply airflow/ti_deps/deps/branch_dep.py]: No I can't find a test for this useful behaviour. So I'm adding one.

File: airflow/ti_deps/deps/branch_dep.py (hunk 5 above)
Comment (yuqian90) [reply airflow/ti_deps/deps/branch_dep.py]: @ash okay this is done. I created ``DepContext.ensure_finished_tasks()`` and made existing code use it.

However, I only found two uses, one in ``TriggerRuleDep`` and one in my ``NotPreviouslySkippedDep``. Where is the new one you mentioned?

**Comment by yuqian90** [comment]
Hi @ashb  thanks for you review. I have adopted all your suggestions, with this tweak:

- I moved the XCom code into ``SkipMixin``. The main use of this ``BranchPythonOperator`` and ``BaseBranchOperator`` both determine tasks to skip by returning branches to **follow**. So I find it important to have ``SkipMixin`` store the branches **followed** rather than the ones skipped when ``skip_all_except()`` is called. When ``skip()`` is called, obviously it should store the tasks that are skipped. So the re-vamped ``SkipMixin`` now handles both cases. The shared code is refactored into ``_set_state_to_skipped()`` to avoid duplication.

Please take another look and let me know what you think.

File: airflow/models/skipmixin.py [hunk 11]
```diff
@@ -16,42 +16,38 @@
...
         if dag_run:
             session.query(TaskInstance).filter(
                 TaskInstance.dag_id == dag_run.dag_id,
                 TaskInstance.execution_date == dag_run.execution_date,
-                TaskInstance.task_id.in_(task_ids)
-            ).update({TaskInstance.state: State.SKIPPED,
-                      TaskInstance.start_date: now,
-                      TaskInstance.end_date: now},
-                     synchronize_session=False)
-            session.commit()
```
Comment (ashb) [inline airflow/models/skipmixin.py:54]: Why did you remove the commit here? (It might be okay, just curious as to your reasoning.)

Oh you renamed the FN to. I see. All good.

File: airflow/models/skipmixin.py [hunk 12]
```diff
@@ -65,13 +61,44 @@ def skip(self, dag_run, execution_date, tasks, session=None):
...
+        :param session: db session to use
+        :param ti: The TaskInstance that initiates the skip. Used for storing results to XCom.
+            If not given, skipped task IDs will not be stored to XCom.
+        """
+        if not tasks:
+            return
+
+        self._set_state_to_skipped(dag_run, execution_date, tasks, session)
+        session.commit()
+        if ti:
```
Comment (ashb) [inline airflow/models/skipmixin.py]: We can avoid needing a TI here, and thus not have to change the fn sigature by calling `XCom.set` directly:

```
XCom.set(
    key=XCOM_SKIPMIXIN_KEY,
    value={"skipped": [d.task_id for d in tasks],
    execution_date=execution_date,
    task_id=self.task_id,
    dag_id=dag_run.dag_id,
    session=session
)
```

File: airflow/ti_deps/deps/not_previously_skipped_dep.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+                    continue
+
+                prev_result = ti.xcom_pull(task_ids=parent.task_id, key=XCOM_SKIPMIXIN_KEY)
+
+                if prev_result is None:
+                    # This can happen if the parent task has not yet run.
+                    continue
+
+                should_skip = False
+                if "followed" in prev_result and ti.task_id not in prev_result["followed"]:
```
Comment (ashb) [inline airflow/ti_deps/deps/not_previously_skipped_dep.py]: Can we make "followed" and "skipped" constants in SkipMixin module please?

File: airflow/ti_deps/deps/not_previously_skipped_dep.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+                    # Skip any tasks that are in "skipped"
+                    should_skip = True
+
+                if should_skip:
+                    # If the parent SkipMixin has run, and the XCom result stored indicates this
+                    # ti should be skipped, set ti.state to SKIPPED and fail the rule so that the
+                    # ti does not execute.
+                    ti.set_state(State.SKIPPED, session)
+                    yield self._failing_status(
+                        reason="Skipping {} because of previous XCom result from "
```
Comment (ashb) [inline airflow/ti_deps/deps/not_previously_skipped_dep.py]: We don't normally include the current task ID in these messages (as they are displayed in the context of a task anyway.

```suggestion
                        reason="Skipping since parent task {} decided this task should be skipped"
```
might do it?

File: tests/ti_deps/deps/test_not_previously_skipped_dep.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    op2 = DummyOperator(task_id="op2", dag=dag)
+    op3 = DummyOperator(task_id="op3", dag=dag)
+    op1 >> [op2, op3]
+
+    TaskInstance(op1, start_date).run()
+    ti2 = TaskInstance(op2, start_date)
+
+    with create_session() as session:
+        dep = NotPreviouslySkippedDep()
+        assert len(list(dep._get_dep_statuses(ti2, session, DepContext()))) == 1
```
Comment (ashb) [inline tests/ti_deps/deps/test_not_previously_skipped_dep.py]: Do we need to test _get_dap_statuses and dep.is_met? That feels like they are both testing exactly the same thing.

And we should test `dep.get_dep_statuses` if we keep both.

**Review by ashb — CHANGES_REQUESTED** [review]
Looking good! A couple of small tweaks/questions, but like where this is heading.

File: tests/ti_deps/deps/test_not_previously_skipped_dep.py (hunk 15 above)
Comment (yuqian90) [reply tests/ti_deps/deps/test_not_previously_skipped_dep.py]: It's testing both because it's possible that the `len()` is more than zero and is_met() is True. It's not happening here, but I think it can happen if the dep yields a `_passing_status()`. i changed this to call `get_dep_statuses()` like you suggested.

File: airflow/models/skipmixin.py (hunk 11 above)
Comment (yuqian90) [reply airflow/models/skipmixin.py:54]: Yes. Thanks. I refactored the shared code into  `_set_state_to_skipped()` and moved the `commit()` into the caller `skip()` and `skip_all_except()`.

File: airflow/models/skipmixin.py (hunk 12 above)
Comment (yuqian90) [reply airflow/models/skipmixin.py]: Oh good idea. One thing i realised though is that `SkipMixin` is not necessarily a `BaseOperator`. So it may not have a `self.task_id` attribute. However, I scanned through the code base I think that only happened in tests. All real use of `SkipMixin` are creating `BaseOperator` from it. Nevertheless I added a type check to make sure this does not break existing code/tests and only do XCom.set() if `self.task_id` is valid.

**Review by yuqian90 — COMMENTED** [review]
Thanks @ashb for the review. I updated the code accordingly.

File: airflow/models/skipmixin.py [hunk 16]
```diff
@@ -88,13 +131,23 @@ def skip_all_except(self, ti: TaskInstance, branch_task_ids: Union[str, Iterable
...
 
             self.log.info("Skipping tasks %s", [t.task_id for t in skip_tasks])
-            self.skip(dag_run, ti.execution_date, skip_tasks)
+            with create_session() as session:
+                self._set_state_to_skipped(
+                    dag_run, ti.execution_date, skip_tasks, session=session
+                )
+                session.commit()
+                ti.xcom_push(
+                    key=XCOM_SKIPMIXIN_KEY, value={"followed": branch_task_ids}
```
Comment (ashb) [inline airflow/models/skipmixin.py]:
```suggestion
                    key=XCOM_SKIPMIXIN_KEY, value={XCOM_SKIPMIXIN_FOLLOWED: branch_task_ids}
```

File: tests/ti_deps/deps/test_trigger_rule_dep.py [hunk 17]
```diff
@@ -19,6 +19,8 @@
 import unittest
 from datetime import datetime
 
+import mock
```
Comment (ashb) [inline tests/ti_deps/deps/test_trigger_rule_dep.py]:
```suggestion
from unittest.mock import Mock
```

and change the usage. (Py3 includes mock, we should move away from using a the separate module as it's not needed)

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by yuqian90** [comment]
Hi, @ashb  thanks for pointing out. I've fixed those. 

Travis is still unhappy, but I'm pretty sure this is unrelated to my change:

File: airflow/models/skipmixin.py [hunk 18]
```diff
@@ -65,13 +68,53 @@ def skip(self, dag_run, execution_date, tasks, session=None):
...
                 session.merge(ti)
 
-            session.commit()
+    @provide_session
+    def skip(
+        self, dag_run, execution_date, tasks, session=None,
+    ):
+        """
+        Sets tasks instances to skipped from the same dag run.
+        If `task_id` is a valid attribute, store the list of skipped task IDs to XCom
```
Comment (ashb) [inline airflow/models/skipmixin.py]:
```suggestion

        If this instance has a `task_id` attribute, store the list of skipped task IDs to XCom
```

File: airflow/models/skipmixin.py [hunk 19]
```diff
@@ -65,13 +68,53 @@ def skip(self, dag_run, execution_date, tasks, session=None):
...
+
+        if task_id is not None:
+            from airflow.models.xcom import XCom
+
+            XCom.set(
+                key=XCOM_SKIPMIXIN_KEY,
+                value={XCOM_SKIPMIXIN_SKIPPED: [d.task_id for d in tasks]},
+                task_id=task_id,
+                dag_id=dag_run.dag_id,
+                execution_date=dag_run.execution_date,
```
Comment (ashb) [inline airflow/models/skipmixin.py:107]:
```suggestion
                execution_date=dag_run.execution_date,
                session=session
```

File: airflow/models/skipmixin.py [hunk 20]
```diff
@@ -65,13 +68,53 @@ def skip(self, dag_run, execution_date, tasks, session=None):
...
 
-    def skip_all_except(self, ti: TaskInstance, branch_task_ids: Union[str, Iterable[str]]):
+    def skip_all_except(
+        self, ti: TaskInstance, branch_task_ids: Union[str, Iterable[str]]
+    ):
         """
         This method implements the logic for a branching operator; given a single
         task ID or list of task IDs to follow, this skips all other tasks
-        immediately downstream of this operator.
+        immediately downstream of this operator. branch_task_ids is stored to XCom so that
```
Comment (ashb) [inline airflow/models/skipmixin.py]:
```suggestion
        immediately downstream of this operator.
        
        branch_task_ids is stored to XCom so that
```

(Plus reformat.)

File: airflow/models/skipmixin.py (hunk 19 above)
Comment (yuqian90) [reply airflow/models/skipmixin.py:107]: Good catch!

File: airflow/models/skipmixin.py [hunk 21]
```diff
@@ -88,13 +134,23 @@ def skip_all_except(self, ti: TaskInstance, branch_task_ids: Union[str, Iterable
...
+                and t.task_id not in branch_downstream_task_ids
+            ]
 
             self.log.info("Skipping tasks %s", [t.task_id for t in skip_tasks])
-            self.skip(dag_run, ti.execution_date, skip_tasks)
+            with create_session() as session:
+                self._set_state_to_skipped(
+                    dag_run, ti.execution_date, skip_tasks, session=session
+                )
+                session.commit()
```
Comment (ashb) [inline airflow/models/skipmixin.py]:
```suggestion
```

Not needed, the session will be committed automatically by `create_session` (and not explicitly committing makes the xcom and the status update be a single transaction.

**Review by ashb — APPROVED** [review]
